pkg/dispers/conceptindexor.go

```
@@ -0,0 +1,198 @@
+
+	// Create index to run a mango query
+	hash := ""
+
+	// Precise and run the mango query
+	var out []ConceptDoc
+	req := &couchdb.FindRequest{Selector: mango.Equal("concept", concept)}
+	err := couchdb.FindDocs(prefixerCI, doctypeSalt, req, &out)
+	if err != nil {
+		return hash, err
+	}
+
+	// Create a salt if no salt in the database for this concept
+	if len(out) == 1 {
+		hash = out[0].Hash
+	} else if len(out) == 0 {
+		return "", errors.New("Concept Indexor : no existing salt for this concept")
+	} else {
+		return "", errors.New("Concept Indexor : several salts for this concept")
+	}
+
+	return hash, err
+}
+
+func createHash(str string) (string, error) {
+
+	salt := string(crypto.GenerateRandomBytes(defaultLenSalt))
+
+	h := sha256.New()
+	h.Write([]byte(str + salt))
+	hash := h.Sum([]byte("concept"))
+
+	return string(hash), nil
+}
+
+func isConceptExisting(concept string) (bool, error) {
+
+	// Precise and run the mango query
+	var out []ConceptDoc
+	req := &couchdb.FindRequest{Selector: mango.Equal("concept", concept)}
+	err := couchdb.FindDocs(prefixerCI, doctypeSalt, req, &out)
+	if err != nil {
+		return false, err
+	}
+
+	if len(out) > 0 {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// CreateConcept checks if concept exists in db. If yes, return error. If no, create the salt, save the concept in db and return the hash.
+func CreateConcept(in *dispers.Concept) error {
+
+	if err := couchdb.EnsureDBExist(prefixerCI, doctypeSalt); err != nil {
+		return err
+	}
+
+	// Get salt with concept
+	isExisting, err := isConceptExisting(in.Concept)
+	if err != nil {
+		return err
+	}
+
+	if isExisting {
+		return errors.New("Concept is already existing")
+	}
+	if err = saveConcept(in.Concept); err != nil {
+		return err
+	}
+	in.Hash, err = getHash(in.Concept)
+
+	return err
+}
+
+// GetConcept gets a concept from db. If no, return error.
+func GetConcept(in *dispers.Concept) error {
+
+	err := couchdb.EnsureDBExist(prefixerCI, doctypeSalt)
```

The db is already created at insertion ; here you can simply return the couchdb error, which will be [this one](https://github.com/cozy/cozy-stack/blob/745bb33ab8163c0d7dab7525068a3b4b48ad9382/pkg/couchdb/errors.go#L108) in the db does not exist.